# CNN 
Assume user downloaded archive.zip from Kaggle, renamed the file BuildingData.zip, and stored the file in the data subdirectory. Assume the zip file contains the weather.csv file.

Use neural net with CNN layers to predict steam usage based on weather.

The model was trained for 25 epochs.
The model was trained on Year 1 and tested on Year 2.
Only selected buildings of site Eagle wered used.

The predictor variables are the 8 features of weather
over the previous 24 time periods.
The predicted variable is the 1 steam usage feature
over the future 1 time period.

In [1]:
DATAPATH=''
try:
    # On Google Drive, set path to my drive / data directory.
    from google.colab import drive
    IN_COLAB = True
    PATH='/content/drive/'
    drive.mount(PATH)
    DATAPATH=PATH+'My Drive/data/'  # must end in "/"
except:
    # On home computer, set path to local data directory.
    IN_COLAB = False
    DATAPATH='data/'  # must end in "/"

ZIP_FILE='BuildingData.zip'
ZIP_PATH = DATAPATH+ZIP_FILE
STEAM_FILE='steam.csv'
WEATHER_FILE='weather.csv'
MODEL_FILE='Model'  # will be used later to save models

In [2]:
from os import listdir
import csv
from zipfile import ZipFile
import numpy as np
import pandas as pd
from scipy import stats  # mode

from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error

from keras.models import Sequential
from keras.layers import SimpleRNN
from keras.layers import LSTM
from keras.layers import TimeDistributed
from keras.layers import Dense
from keras.losses import MeanSquaredError
from keras.layers import Conv2D
from keras.layers import Flatten

import matplotlib.pyplot as plt
from matplotlib import colors
mycmap = colors.ListedColormap(['red','blue'])  # list color for label 0 then 1
np.set_printoptions(precision=2)

In [3]:
def read_zip_to_panda(zip_filename,csv_filename):
    zip_handle = ZipFile(zip_filename)
    csv_handle = zip_handle.open(csv_filename)
    panda = pd.read_csv(csv_handle)
    return panda
def fix_date_type(panda):
    # Convert the given timestamp column to the pandas datetime data type.
    panda['timestamp'] = pd.to_datetime(panda['timestamp'], infer_datetime_format = True)
    indexed = panda.set_index(['timestamp'])
    return indexed
def get_site_timeseries(panda,site):
    # Assume the panda dataframe has a datetime column.
    # (If not, call fix_date_type() before this.)
    # Extract the timeseries for one site.
    # Convert the datetime column to a DatetimeIndex.
    site_df = panda[panda['site_id']==site]
    temp_col = site_df['date']
    temp_val = temp_col.values
    temp_ndx = pd.DatetimeIndex(temp_val)
    dropped = site_df.drop('date',axis=1)
    panda = dropped.set_index(temp_ndx)
    return panda

## CNN setup

In [4]:
# Before analyzing the entire dataset, we look at this subset.
SITE = 'Eagle'
METER = 'steam'

# Arrange "picture" of weather with temperatures toward the middle
PREDICTED_VARIABLE = 'steam' 
PREDICTORS = ['cloudCoverage', 'airTemperature', 'dewTemperature', 'precipDepth1HR', 'precipDepth6HR', 'seaLvlPressure', 'windDirection', 'windSpeed']
print("PREDICTORS=",len(PREDICTORS),PREDICTORS)

# Downsample True means collapse 365*24 measures to 365 daily averages
# Downsample False means replace 365*24 measures with 365*24 window averages
DOWNSAMPLE = False   

STEPS_HISTORY = 24   # length of the predictor sequence
STEPS_FUTURE =  1    # length of the predicted sequence

## CNN parameters
EPOCHS=25
FILTERS = 8
WIDTH = 3
STRIDE = (1,1)
INPUT_SHAPE = (STEPS_HISTORY,len(PREDICTORS),1) 

PREDICTORS= 8 ['cloudCoverage', 'airTemperature', 'dewTemperature', 'precipDepth1HR', 'precipDepth6HR', 'seaLvlPressure', 'windDirection', 'windSpeed']


In [5]:
wet_df = read_zip_to_panda(ZIP_PATH,WEATHER_FILE)
wet_df = fix_date_type(wet_df)
stm_df = read_zip_to_panda(ZIP_PATH,STEAM_FILE)
stm_df = fix_date_type(stm_df)
site_specific_weather = wet_df.loc[wet_df['site_id'] == SITE]
all_buildings = [x for x in stm_df.columns if x.startswith(SITE)]

In [6]:
def smooth(df):
    # For smoothing the 24 hour cycle, we do not want exponential smoothing.
    smoothed = None
    if DOWNSAMPLE:
        # This alternate method samples down to 1/24 time steps.
        smoothed = df.resample("24H").mean() 
    else:
        # This method does not reduce the number of time steps.
        # Note the first 23 measurements get set to Nan.
        smoothed=df.rolling(window=24).mean()
        smoothed=smoothed[24:]
    return smoothed

# Correlation is low when buildings have many NaN and 0 meter readings.
# We will ignore buildings that have >max bad meter readings.
def is_usable_column(df,column_name):
    MAX_BAD = 500 
    bad = df[column_name].isin([0]).sum()
    return bad<=MAX_BAD

def prepare_for_learning(df):
    num_samples = len(df) - STEPS_FUTURE - STEPS_HISTORY
    num_predictors = len(PREDICTORS)
    X_shape = (num_samples,STEPS_HISTORY,num_predictors,1)
    X=np.zeros(X_shape)
    Y_shape = (num_samples,STEPS_FUTURE)
    y=np.zeros(Y_shape)
    predictor_series = df[PREDICTORS].values  # e.g. all weather values
    predicted_series = df[PREDICTED_VARIABLE].values  # e.g. all meter readings
    
    for x0 in range (0,num_samples): # Loop over all 1000 samples
        # This is one array of weather for previous 24 time periods
        one_sample = predictor_series[x0:x0+STEPS_HISTORY]
        one_label =  predicted_series[x0+STEPS_HISTORY:x0+STEPS_FUTURE]
        # Loop over all 24 time periods
        for x1 in range (0,STEPS_HISTORY): # In 1 sample, loop over 24 time periods
            one_period = one_sample[x1]
            for x2 in range (0,num_predictors): # In 1 time period, loop over 8 weather metrics
                one_predictor = one_period[x2]
                # for x3 in range (0,X_shape[3]): # In 1 metric, loop over vector dimensions
                # In our data, each weather metric is a scalar.
                x3 = 0
                X[x0,x1,x2,x3] = one_predictor
        y[x0]=predicted_series[x0:x0+STEPS_FUTURE]
    return X,y 


In [7]:
def make_CNN():
    print("make_CNN")
    print("input shape:",INPUT_SHAPE)
    cnn = Sequential()
    cnn.add(
        Conv2D( input_shape=INPUT_SHAPE,
            filters=FILTERS,kernel_size=WIDTH,strides=STRIDE,
            activation=None, padding="valid"))
    cnn.add(Flatten())
    cnn.add(Dense(STEPS_FUTURE))   
    cnn.compile(optimizer='adam',loss=MeanSquaredError())
    return cnn

## Process all buildings

In [8]:
cors = []
ONE_PREDICTOR = 'dewTemperature'  ## illustrate difficulty by showing best correlate
for BLDG in all_buildings:
    print("Building",BLDG)
    # Get steam usage for one building.
    bldg_specific_steam = stm_df[[BLDG]]
    # Concatenate steam usage with weather.
    one_bldg_df = pd.concat([bldg_specific_steam,site_specific_weather],axis=1)
    # Drop the site, which is constant (we selected for one site).
    one_bldg_df = one_bldg_df.drop(['site_id'],axis=1)
    # The original steam table used column name = building name.
    # We are processing one building, so rename to the column 'steam'.
    one_bldg_df = one_bldg_df.rename(columns={BLDG : METER})
    # In order to filter bad buildings, count sum of NaN + zero.
    one_bldg_df = one_bldg_df.fillna(0)
    
    if is_usable_column(one_bldg_df,METER):
        one_bldg_df = smooth(one_bldg_df) 
        X,y = prepare_for_learning(one_bldg_df)
        # Ideally, split Year1 = train, Year2 = test.
        # Some data is incomplete, so split 1st half and 2nd half.
        split = len(X)//2 
        X_train = np.asarray(X[0:split])
        y_train = np.asarray(y[0:split])
        X_test = np.asarray(X[split:])
        y_test = np.asarray(y[split:])

        model = make_CNN()
        print(model.summary())
        print("X_train.shape:",X_train.shape)
        model.fit(X_train,y_train,epochs=EPOCHS)
        y_pred = model.predict(X_test)
        rmse = mean_squared_error(y_test,y_pred,squared=False)
        # Keep a table for reporting later.
        mean = one_bldg_df[METER].mean()
        cor = one_bldg_df.corr().loc[PREDICTED_VARIABLE][ONE_PREDICTOR] 
        cors.append([cor,mean,rmse,rmse/mean,BLDG])
        print("cor,mean,rmse,rmse/mean,bldg:",cor,mean,rmse,rmse/mean,BLDG)
        
print()
print("History",STEPS_HISTORY,"Future",STEPS_FUTURE)
print("Column 1: Correlation of",PREDICTED_VARIABLE,"and",ONE_PREDICTOR)
print("          Using one weather feature as leading correlate.")
print("Column 2: Mean usage.")
print("          Using mean to help understand the RMSE.")
print("Column 3: RMSE of LinearRegression(X=Weather, y=Usage).")
print("Column 4: RMSE/mean normalized to help understand RMSE.")
print("Column 5: Building.")
for cor in sorted(cors):
    print("%7.4f %10.2f %10.2f %5.2f   %s"%(cor[0],cor[1],cor[2],cor[3],cor[4]))    

Building Eagle_office_Lamont
Building Eagle_health_Athena
make_CNN
input shape: (24, 8, 1)
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 22, 6, 8)          80        
_________________________________________________________________
flatten (Flatten)            (None, 1056)              0         
_________________________________________________________________
dense (Dense)                (None, 1)                 1057      
Total params: 1,137
Trainable params: 1,137
Non-trainable params: 0
_________________________________________________________________
None
X_train.shape: (8747, 24, 8, 1)
Epoch 1/25
274/274 [==============================] - 1s 3ms/step - loss: 45037.2703
Epoch 2/25
274/274 [==============================] - 1s 3ms/step - loss: 20569.9467
Epoch 3/25
274/274 [==============================] - 1s 2ms/step - loss: 19623.3970
E

274/274 [==============================] - 1s 4ms/step - loss: 76948.5809
Epoch 14/25
274/274 [==============================] - 1s 4ms/step - loss: 78516.8076
Epoch 15/25
274/274 [==============================] - 1s 4ms/step - loss: 76916.2447
Epoch 16/25
274/274 [==============================] - 1s 4ms/step - loss: 73395.4756
Epoch 17/25
274/274 [==============================] - 1s 4ms/step - loss: 78565.4549
Epoch 18/25
274/274 [==============================] - 1s 4ms/step - loss: 75386.2971
Epoch 19/25
274/274 [==============================] - 1s 4ms/step - loss: 74269.2277
Epoch 20/25
274/274 [==============================] - 1s 4ms/step - loss: 73932.9071
Epoch 21/25
274/274 [==============================] - 1s 4ms/step - loss: 71646.7821
Epoch 22/25
274/274 [==============================] - 1s 4ms/step - loss: 74351.9345
Epoch 23/25
274/274 [==============================] - 1s 4ms/step - loss: 73099.2258
Epoch 24/25
274/274 [==============================] - 1s 4ms/step

274/274 [==============================] - 1s 2ms/step - loss: 9705.4227
Epoch 2/25
274/274 [==============================] - 1s 2ms/step - loss: 1467.3110
Epoch 3/25
274/274 [==============================] - 1s 2ms/step - loss: 1380.7466
Epoch 4/25
274/274 [==============================] - 1s 2ms/step - loss: 1370.2668
Epoch 5/25
274/274 [==============================] - 1s 2ms/step - loss: 1308.8847
Epoch 6/25
274/274 [==============================] - 1s 2ms/step - loss: 1352.8173
Epoch 7/25
274/274 [==============================] - 1s 2ms/step - loss: 1359.8797
Epoch 8/25
274/274 [==============================] - 1s 2ms/step - loss: 1367.7725
Epoch 9/25
274/274 [==============================] - 1s 2ms/step - loss: 1387.3196
Epoch 10/25
274/274 [==============================] - 1s 2ms/step - loss: 1255.0040
Epoch 11/25
274/274 [==============================] - 1s 2ms/step - loss: 1310.0415
Epoch 12/25
274/274 [==============================] - 1s 2ms/step - loss: 1294.5992


274/274 [==============================] - 1s 2ms/step - loss: 519.0592
Epoch 25/25
274/274 [==============================] - 1s 2ms/step - loss: 534.7642
cor,mean,rmse,rmse/mean,bldg: -0.7961450049576672 121.90840096508735 31.107638323241716 0.2551722283040236 Eagle_health_Vincenza
Building Eagle_office_Dallas
Building Eagle_education_Shante
Building Eagle_office_Chauncey
Building Eagle_office_Phyllis
Building Eagle_office_Freida
Building Eagle_office_Francis
make_CNN
input shape: (24, 8, 1)
Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_8 (Conv2D)            (None, 22, 6, 8)          80        
_________________________________________________________________
flatten_8 (Flatten)          (None, 1056)              0         
_________________________________________________________________
dense_8 (Dense)              (None, 1)                 1057      
Total params: 1,1

274/274 [==============================] - 1s 2ms/step - loss: 217732.8090
Epoch 9/25
274/274 [==============================] - 1s 2ms/step - loss: 217328.6714
Epoch 10/25
274/274 [==============================] - 1s 2ms/step - loss: 214305.3146
Epoch 11/25
274/274 [==============================] - 1s 2ms/step - loss: 207428.5752
Epoch 12/25
274/274 [==============================] - 1s 2ms/step - loss: 204415.5768
Epoch 13/25
274/274 [==============================] - 1s 2ms/step - loss: 208484.9344
Epoch 14/25
274/274 [==============================] - 1s 2ms/step - loss: 209570.6254
Epoch 15/25
274/274 [==============================] - 1s 2ms/step - loss: 197370.3428
Epoch 16/25
274/274 [==============================] - 1s 2ms/step - loss: 199561.6153
Epoch 17/25
274/274 [==============================] - 1s 2ms/step - loss: 194564.8739
Epoch 18/25
274/274 [==============================] - 1s 2ms/step - loss: 202273.1933
Epoch 19/25
274/274 [==============================] - 1

274/274 [==============================] - 1s 4ms/step - loss: 3671.7047
Epoch 2/25
274/274 [==============================] - 1s 4ms/step - loss: 431.2885
Epoch 3/25
274/274 [==============================] - 1s 4ms/step - loss: 422.6019
Epoch 4/25
274/274 [==============================] - 1s 4ms/step - loss: 421.6188
Epoch 5/25
274/274 [==============================] - 1s 4ms/step - loss: 423.9133
Epoch 6/25
274/274 [==============================] - 1s 4ms/step - loss: 398.0669
Epoch 7/25
274/274 [==============================] - 1s 4ms/step - loss: 402.4178
Epoch 8/25
274/274 [==============================] - 1s 4ms/step - loss: 375.2050
Epoch 9/25
274/274 [==============================] - 1s 4ms/step - loss: 360.9399
Epoch 10/25
274/274 [==============================] - 1s 4ms/step - loss: 382.4806
Epoch 11/25
274/274 [==============================] - 1s 4ms/step - loss: 394.3225
Epoch 12/25
274/274 [==============================] - 1s 4ms/step - loss: 363.6223
Epoch 13/25

274/274 [==============================] - 1s 4ms/step - loss: 2592.2553
Epoch 25/25
274/274 [==============================] - 1s 4ms/step - loss: 2546.7516
cor,mean,rmse,rmse/mean,bldg: -0.6796379179294123 148.50737262033823 58.498300069376924 0.39390839011695994 Eagle_education_Teresa
Building Eagle_office_Norbert
Building Eagle_lodging_Casey
Building Eagle_office_Tia
Building Eagle_office_Remedios
Building Eagle_office_Patrice
Building Eagle_education_Shana

History 24 Future 1
Column 1: Correlation of steam and dewTemperature
          Using one weather feature as leading correlate.
Column 2: Mean usage.
          Using mean to help understand the RMSE.
Column 3: RMSE of LinearRegression(X=Weather, y=Usage).
Column 4: RMSE/mean normalized to help understand RMSE.
Column 5: Building.
-0.8877    2030.36     402.31  0.20   Eagle_education_Sherrill
-0.8545    1634.28     649.23  0.40   Eagle_education_Brooke
-0.8502    3147.43     729.32  0.23   Eagle_education_Peter
-0.8403     477.4

### Report 1
Report 1, Table I, includes the following summary. This is the mean over 16 builings of the normalized RMSE per building.

CNN model using predictions based on 24 times, 8 features

* 0.85 mean RMSE
* 2.07 stddev

Here are the results omitting outlier building Wesley.

* 0.33 mean RMSE
* 0.09 stddev


### Report 2
We reran the notebook and got slightly different results.

* 0.74 mean RMSE
* 1.61 stddev

Here are the results omitting outlier building Wesley.

* 0.34 mean RMSE
* 0.09 stddev
